File: copilot-connector/sample_codes/java/common-patterns/ThrottleResilientIngestion.java
// Throttle-resilient ingestion with batch concurrency control
// Shows: retry with exponential backoff on 429, batch ingestion with concurrency limits, auto-chunking
//
// Prerequisites:
//   Maven: com.microsoft.graph:microsoft-graph:6.x, com.azure:azure-identity:1.x
//   Gradle: implementation 'com.microsoft.graph:microsoft-graph:6.+'
//           implementation 'com.azure:azure-identity:1.+'

import com.microsoft.graph.GraphServiceClient;
import com.microsoft.graph.models.externalconnectors.*;
import com.microsoft.graph.models.externalconnectors.Properties;
import com.microsoft.kiota.ApiException;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.BiConsumer;

/**
 * Demonstrates throttle-resilient ingestion strategies for a Copilot Connector.
 * Includes exponential backoff retry, batch concurrency control via Semaphore,
 * and automatic content chunking for items exceeding the 4 MB limit.
 */
public class ThrottleResilientIngestion {

    private final GraphServiceClient graphClient;
    private final String connectionId;

    public ThrottleResilientIngestion(GraphServiceClient graphClient, String connectionId) {
        this.graphClient = graphClient;
        this.connectionId = connectionId;
    }

    /**
     * Ingest a single item with exponential backoff retry on throttling (HTTP 429).
     *
     * @param item       the external item to ingest
     * @param maxRetries maximum number of retry attempts
     */
    public void ingestWithRetry(ExternalItem item, int maxRetries) throws InterruptedException {
        for (int attempt = 0; attempt <= maxRetries; attempt++) {
            try {
                graphClient.external().connections()
                        .byExternalConnectionId(connectionId)
                        .items()
                        .byExternalItemId(item.getId())
                        .put(item);
                return; // Success
            } catch (ApiException ex) {
                if (ex.getResponseStatusCode() == 429) {
                    if (attempt == maxRetries) {
                        throw ex;
                    }

                    // Use Retry-After header if available, otherwise exponential backoff
                    long retryAfterSeconds = parseRetryAfter(ex, attempt);

                    System.out.printf("Throttled on item %s. Retry %d/%d after %ds%n",
                            item.getId(), attempt + 1, maxRetries, retryAfterSeconds);

                    Thread.sleep(retryAfterSeconds * 1000);
                } else {
                    throw ex; // Non-throttling error — don't retry
                }
            }
        }
    }

    /** Convenience overload with default 5 retries. */
    public void ingestWithRetry(ExternalItem item) throws InterruptedException {
        ingestWithRetry(item, 5);
    }

    /**
     * Batch ingest items with controlled concurrency.
     * Uses a Semaphore to limit parallel requests (recommended: 4-8,
     * never exceeding the 25 concurrent operations per connection limit).
     *
     * @param items          items to ingest
     * @param maxConcurrency maximum parallel ingestion requests
     * @param progress       optional callback receiving (completed, total) counts
     */
    public void batchIngest(
            List<ExternalItem> items,
            int maxConcurrency,
            BiConsumer<Integer, Integer> progress) throws InterruptedException {

        var semaphore = new Semaphore(maxConcurrency);
        var completed = new AtomicInteger(0);
        int total = items.size();

        try (var executor = Executors.newVirtualThreadPerTaskExecutor()) {
            var futures = new ArrayList<Future<?>>();

            for (var item : items) {
                Future<?> future = executor.submit(() -> {
                    try {
                        semaphore.acquire();
                        try {
                            ingestWithRetry(item);
                            int count = completed.incrementAndGet();
                            if (progress != null) {
                                progress.accept(count, total);
                            }
                        } finally {
                            semaphore.release();
                        }
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        throw new RuntimeException("Ingestion interrupted", e);
                    }
                });
                futures.add(future);
            }

            // Wait for all tasks to complete
            for (var future : futures) {
                try {
                    future.get();
                } catch (ExecutionException e) {
                    System.err.println("Item ingestion failed: " + e.getCause().getMessage());
                }
            }
        }
    }

    /** Convenience overload with default concurrency of 4 and no progress callback. */
    public void batchIngest(List<ExternalItem> items) throws InterruptedException {
        batchIngest(items, 4, null);
    }

    /**
     * Check payload size before ingestion and auto-chunk if needed.
     * The 4 MB limit applies to the full serialized request body.
     *
     * @param item the external item to ingest (may be chunked)
     */
    public void ingestWithAutoChunk(ExternalItem item) throws InterruptedException {
        // Estimate serialized size from content + properties
        String contentValue = item.getContent() != null ? item.getContent().getValue() : "";
        int byteSize = contentValue.getBytes(StandardCharsets.UTF_8).length;

        if (byteSize > 3_800_000) { // Leave 200KB buffer below the 4 MB limit
            System.out.printf("Item %s is %.1fMB — chunking required.%n",
                    item.getId(), byteSize / 1_000_000.0);

            List<String> chunks = chunkContent(contentValue, 3_500_000);

            for (int i = 0; i < chunks.size(); i++) {
                ExternalItem chunkedItem = cloneItemWithChunk(item, chunks.get(i), i, chunks.size());
                ingestWithRetry(chunkedItem);
            }
        } else {
            ingestWithRetry(item);
        }
    }

    /**
     * Split content at paragraph boundaries, respecting byte size limits.
     */
    private static List<String> chunkContent(String content, int maxChunkBytes) {
        String[] paragraphs = content.split("\n\n");
        var chunks = new ArrayList<String>();
        var current = new StringBuilder();
        int currentSize = 0;

        for (String para : paragraphs) {
            if (para.isBlank()) continue;

            int paraSize = para.getBytes(StandardCharsets.UTF_8).length;

            if (currentSize + paraSize > maxChunkBytes && !current.isEmpty()) {
                chunks.add(current.toString());
                current.setLength(0);
                currentSize = 0;
            }

            current.append(para).append("\n\n");
            currentSize += paraSize;
        }

        if (!current.isEmpty()) {
            chunks.add(current.toString());
        }

        return chunks;
    }

    /**
     * Create a chunked version of an item with a unique ID and title context.
     */
    private static ExternalItem cloneItemWithChunk(
            ExternalItem original, String chunkContent, int chunkIndex, int totalChunks) {

        String originalTitle = "Document";
        if (original.getProperties() != null
                && original.getProperties().getAdditionalData() != null
                && original.getProperties().getAdditionalData().containsKey("title")) {
            originalTitle = original.getProperties().getAdditionalData().get("title").toString();
        }

        var props = new HashMap<String, Object>();
        if (original.getProperties() != null && original.getProperties().getAdditionalData() != null) {
            props.putAll(original.getProperties().getAdditionalData());
        }
        props.put("title", originalTitle + " (Part " + (chunkIndex + 1) + " of " + totalChunks + ")");
        props.put("parentDocumentId", original.getId() != null ? original.getId() : "");
        props.put("chunkIndex", chunkIndex);
        props.put("totalChunks", totalChunks);

        // Prepend contextual header for self-contained chunks
        String contextHeader = "Document: " + originalTitle + "\n"
                + "Part " + (chunkIndex + 1) + " of " + totalChunks + "\n\n";

        ExternalItem chunkedItem = new ExternalItem();
        chunkedItem.setId(original.getId() + "_chunk_" + chunkIndex);
        chunkedItem.setAcl(original.getAcl());

        Properties properties = new Properties();
        properties.setAdditionalData(props);
        chunkedItem.setProperties(properties);

        ExternalItemContent content = new ExternalItemContent();
        content.setValue(contextHeader + chunkContent);
        content.setType(original.getContent() != null
                ? original.getContent().getType()
                : ExternalItemContentType.Text);
        chunkedItem.setContent(content);

        return chunkedItem;
    }

    /**
     * Parse Retry-After from the exception, falling back to exponential backoff.
     */
    private static long parseRetryAfter(ApiException ex, int attempt) {
        // The Kiota ApiException may carry response headers; try to extract Retry-After.
        // Fall back to exponential backoff: 1s, 2s, 4s, 8s, 16s...
        try {
            var headers = ex.getResponseHeaders();
            if (headers != null) {
                var retryValues = headers.get("Retry-After");
                if (retryValues != null && !retryValues.isEmpty()) {
                    return Long.parseLong(retryValues.iterator().next());
                }
            }
        } catch (Exception ignored) {
            // Header parsing failed — use fallback
        }
        return (long) Math.pow(2, attempt);
    }
}

File: copilot-connector/sample_codes/python/common-patterns/throttle-resilient-ingestion.py
# Throttle-resilient ingestion with batch concurrency control
# Prerequisites: pip install msgraph-sdk azure-identity

import asyncio
import json
from typing import Callable, Optional

from msgraph import GraphServiceClient
from msgraph.generated.models.external_connectors.external_item import ExternalItem
from msgraph.generated.models.external_connectors.external_item_content import ExternalItemContent
from msgraph.generated.models.external_connectors.external_item_content_type import ExternalItemContentType
from msgraph.generated.models.external_connectors.properties import Properties
from msgraph.generated.models.o_data_errors.o_data_error import ODataError


class ResilientIngestion:
    """Manages throttle-resilient item ingestion with concurrency control."""

    def __init__(self, graph_client: GraphServiceClient, connection_id: str) -> None:
        self._graph_client = graph_client
        self._connection_id = connection_id

    async def ingest_with_retry(
        self, item: ExternalItem, max_retries: int = 5
    ) -> None:
        """Ingest a single item with exponential backoff retry on throttling."""
        for attempt in range(max_retries + 1):
            try:
                await self._graph_client.external.connections.by_external_connection_id(
                    self._connection_id
                ).items.by_external_item_id(item.id).put(item)
                return  # Success
            except ODataError as e:
                if e.response_status_code != 429:
                    raise

                if attempt == max_retries:
                    raise

                # Respect Retry-After header if available, otherwise exponential backoff
                retry_after = 2**attempt
                if e.response_headers:
                    retry_values = e.response_headers.get("Retry-After", [])
                    if retry_values:
                        try:
                            retry_after = int(retry_values[0])
                        except (ValueError, IndexError):
                            pass

                print(
                    f"Throttled on item {item.id}. "
                    f"Retry {attempt + 1}/{max_retries} after {retry_after}s"
                )
                await asyncio.sleep(retry_after)

    async def batch_ingest(
        self,
        items: list[ExternalItem],
        max_concurrency: int = 4,
        on_progress: Optional[Callable[[int, int], None]] = None,
    ) -> None:
        """
        Batch ingest items with controlled concurrency.
        Uses a semaphore to limit parallel requests (recommended: 4-8,
        never exceeding the 25 concurrent operations per connection limit).
        """
        semaphore = asyncio.Semaphore(max_concurrency)
        completed = 0

        async def _ingest_one(item: ExternalItem) -> None:
            nonlocal completed
            async with semaphore:
                await self.ingest_with_retry(item)
                completed += 1
                if on_progress:
                    on_progress(completed, len(items))

        await asyncio.gather(*[_ingest_one(item) for item in items])

    async def ingest_with_auto_chunk(self, item: ExternalItem) -> None:
        """
        Check payload size before ingestion and auto-chunk if needed.
        The 4 MB limit applies to the full serialized request body.
        """
        byte_size = _estimate_payload_size(item)

        if byte_size > 3_800_000:  # Leave 200KB buffer below 4MB limit
            print(
                f"Item {item.id} is {byte_size / 1_000_000:.1f}MB — chunking required."
            )

            chunks = _chunk_content(
                item.content.value if item.content else "",
                max_chunk_bytes=3_500_000,
            )

            for i, chunk in enumerate(chunks):
                chunked_item = _clone_item_with_chunk(item, chunk, i, len(chunks))
                await self.ingest_with_retry(chunked_item)
        else:
            await self.ingest_with_retry(item)


def _estimate_payload_size(item: ExternalItem) -> int:
    """Estimate the serialized payload size in bytes."""
    payload: dict = {}
    if item.id:
        payload["id"] = item.id
    if item.properties and item.properties.additional_data:
        payload["properties"] = item.properties.additional_data
    if item.content and item.content.value:
        payload["content"] = item.content.value
    return len(json.dumps(payload, default=str).encode("utf-8"))


def _chunk_content(content: str, max_chunk_bytes: int) -> list[str]:
    """Split content at paragraph boundaries, respecting byte size limits."""
    paragraphs = content.split("\n\n")
    chunks: list[str] = []
    current_parts: list[str] = []
    current_size = 0

    for para in paragraphs:
        if not para.strip():
            continue

        para_size = len(para.encode("utf-8"))

        if current_size + para_size > max_chunk_bytes and current_parts:
            chunks.append("\n\n".join(current_parts))
            current_parts = []
            current_size = 0

        current_parts.append(para)
        current_size += para_size

    if current_parts:
        chunks.append("\n\n".join(current_parts))

    return chunks


def _clone_item_with_chunk(
    original: ExternalItem, chunk_content: str, chunk_index: int, total_chunks: int
) -> ExternalItem:
    """Create a chunked version of an item with a unique ID and title context."""
    original_props = (
        dict(original.properties.additional_data)
        if original.properties and original.properties.additional_data
        else {}
    )

    original_title = original_props.get("title", "Document")

    props = dict(original_props)
    props["title"] = f"{original_title} (Part {chunk_index + 1} of {total_chunks})"
    props["parentDocumentId"] = original.id or ""
    props["chunkIndex"] = chunk_index
    props["totalChunks"] = total_chunks

    # Prepend contextual header for self-contained chunks
    context_header = (
        f"Document: {original_title}\n"
        f"Part {chunk_index + 1} of {total_chunks}\n\n"
    )

    return ExternalItem(
        id=f"{original.id}_chunk_{chunk_index}",
        acl=original.acl,
        properties=Properties(additional_data=props),
        content=ExternalItemContent(
            value=context_header + chunk_content,
            type=(
                original.content.type
                if original.content
                else ExternalItemContentType.Text
            ),
        ),
    )
